python/lsst/pipe/base/configOverrides.py

```diff
@@ -125,35 +227,77 @@ def applyTo(self, config):
-                # find object with attribute to set, throws if we name is wrong
-                obj = config
-                for attr in field[:-1]:
-                    obj = getattr(obj, attr)
-                # If input is a string and field type is not a string then we
-                # have to convert string to an expected type. Implementing
-                # full string parser is non-trivial so we take a shortcut here
-                # and `eval` the string and assign the resulting value to a
-                # field. Type erroes can happen during both `eval` and field
-                # assignment.
-                if isinstance(value, str) and obj._fields[field[-1]].dtype is not str:
-                    try:
-                        # use safer ast.literal_eval, it only supports literals
-                        value = ast.literal_eval(value)
-                    except Exception:
-                        # eval failed, wrap exception with more user-friendly
-                        # message
-                        raise pexExceptions.RuntimeError(f"Unable to parse `{value}' into a Python object")
-
-                # this can throw in case of type mismatch
-                setattr(obj, field[-1], value)
+                if isinstance(value, str):
+                    value = self._parser(value, configParser)
+                # checking for dicts and lists here is needed because {} and []
+                # are valid yaml syntax so they get converted before hitting
+                # this method, so we must parse the elements.
+                #
+                # The same override would remain a string if specified on the
+                # command line, and is handled above.
+                if isinstance(value, dict):
+                    new = {}
+                    for k, v in value.items():
+                        if isinstance(v, str):
+                            new[k] = self._parser(v, configParser)
+                        else:
+                            new[k] = v
+                    value = new
+                if isinstance(value, list):
+                    new = []
+                    for v in value:
+                        if isinstance(v, str):
+                            new.append(self._parser(v, configParser))
+                        else:
+                            new.append(v)
+                    value = new
+                # The field might be a string corresponding to a attribute
+                # hierarchy, attempt to split off the last field which
+                # will then be set.
+                parent, *child = field.rsplit(".", maxsplit=1)
+                if child:
+                    # This branch means there was a hierarchy, get the
+                    # field to set, and look up the sub config for which
+                    # it is to be set
+                    finalField = child[0]
+                    tmpConfig = attrgetter(parent)(config)
```

Comment:
Is not this the same as `getattr(config, parent)`?

Reply:
sort of, but with `getattr` you can only get a single attribute from a string. `attrgetter` supports a dotted notation of attributes to fetch; i.e. 'fieldA.fieldB.fieldC.fieldD`. This allows us to go through an arbitrarily deep config structure concisely without loops.

Reply:
Nice, I did not know about that.